### ISSUE TITLE
Q: uBO filter detection

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#362]
Can someone ELI5 how it does this?

@gorhill, only if you have time: not the ELI5 thing, but if it can be blocked globally regardless of scripts, or if this is impossible and I'm making stuff up and misunderstanding the web tech involved

https://browserleaks.com/proxy
![proxy](https://user-images.githubusercontent.com/16656956/36183406-085a70d4-1126-11e8-94fd-779e00461a3c.png)

Note: uBO & uM config = zero 3rd party

Can this be blocked somehow? I need that ELI5 thing

---
After coming across this extension/filter detection in a wilders post, I now see an issue here
- `https://github.com/jspenguin2017/uBlockProtector/issues/808`

and a subsequent filter which the issue discusses - I am not in favor of this filter - all it will do is give a false impression that the info is not leaked in the wild
```
# https://github.com/jspenguin2017/uBlockProtector/issues/808
||browserleaks.com/js/proxy_ab.js$script
```

**Comment by overdodactyl** [comment]
If I recall correctly, these are determined by serving you unique scripts that are found in each of those filter lists.  If the site detects that `x` unique scripts located in filter `y` are being blocked, you are most likely subscribed to it.  

If this is the case, I'm not sure there would be a way to block it..

There was a Reddit post about this a while back, I'll see if I can find it again.

**Comment by Thorin-Oakenpants** [comment]
OK, but I'm blocking every 3rd party request - so why don't other lists get detected - see the pics at jspenguin2017's repo. This is why I need that ELI5 bit - does it have something to do with proxies - how does that work - gimme a flow-chart?

**Comment by Thorin-Oakenpants** [comment]
@overdodactyl  before I forget, whats the bits for userChrome to get rid of the add search section and the update app + symbol

![dactyl-time](https://user-images.githubusercontent.com/16656956/36186794-2fdeeef6-1139-11e8-92e0-1e808e1e6c1f.png)

**Comment by overdodactyl** [comment]
```css
/* Remove add search section */
.search-add-engines {
    display: none;
}
```

Next time I get I FF update I'll check out the little arrow!

**Comment by Thorin-Oakenpants** [comment]
FYI: the same green + icon over the magnifying glass is this
`.searchbar-search-icon-overlay {display: none !important;}`

**Comment by overdodactyl** [comment]
As for detecting the filters, they could serve many of the things found in those lists via first-party scripts, couldn't they?

The EasyList, for example, begins with a long list of "general tracking systems."  A first party (in this case, browserleaks) could just hit you with a bunch of those. The list would only need to contain a few unique items are a unique subset 

https://easylist.to/easylist/easylist.txt

**Comment by Thorin-Oakenpants** [comment]
OK, that makes sense - they re-create unique-per-list global-filters as 1st party. The proxy part on the page threw me for a loop

**Comment by overdodactyl** [comment]
Looks like that might be the case.  New profile, downloaded uBO (no changes) and ran the test with the logger open:

![screen shot](https://user-images.githubusercontent.com/33071129/36188075-804489f2-1141-11e8-9313-fa37906a7565.png)

A first party script (from browserleaks.com) is getting blocked due the first section previously mentioned in EasyPrivacy

**Comment by overdodactyl** [comment]
^ Sorry, didn't see your comment before posting :)

**Comment by Thorin-Oakenpants** [comment]
sheesh man .. I had mine posted at least 24 seconds before yours - look next time dude :trollface:

**Comment by Thorin-Oakenpants** [comment]
Soooo .. FA **we** can do about it then

... although it could be possible (but not feasible) to isolate all the unique stuff into one list enabled by default for all users - way too much work and upkeep, list would get too big, and unless the list handled all the various language lists it would still give away people's localized lists

**Comment by overdodactyl** [comment]
Not sure what a practical approach to preventing that would be..
 
I would have to imagine 90% of uBO users don't ever even touch the settings, so just sticking with the defaults would give up the least amount of info.

**Comment by Thorin-Oakenpants** [comment]
Same goes for CA fingerprinting, although Tor Uplift are looking at FPI for certs [1334485](https://bugzilla.mozilla.org/show_bug.cgi?id=1334485), although that's already handled by blocking 3rd party requests - not really followed the ticket TBH

Same goes for cipher suites, but at least we're at default.

PS: https://cc.dcsec.uni-hannover.de/ - note the top 3 (maybe someone should tell them)

![cipher](https://user-images.githubusercontent.com/16656956/36188982-9fd09530-1147-11e8-9910-fc015af4743c.png)

https://www.howsmyssl.com/
![cipher2](https://user-images.githubusercontent.com/16656956/36189017-c93d8784-1147-11e8-86e9-926e6d84abe2.png)

**Comment by overdodactyl** [comment]
Interesting, hadn't seen anything about CA or cipher suite fingerprinting before.  It is odd three are showing up as unknown (same result here).  I did a quick bugzilla search and didn't see anything about it

**Comment by Thorin-Oakenpants** [comment]
> It is odd three are showing up as unknown

because the site hasn't coded for them is my guess

As for CA and cipher FP'ing - been discussed for ages - there's a CA test one in the wiki test links. You could also do a FP on media types the browser handles (eg see the youtube html5 test page - https://www.youtube.com/html5: imagine something like that with mp3, ogg, flac, webm, etc).

**Comment by Thorin-Oakenpants** [comment]
CA cache (fishing for unique CA's etc), easily defeated

https://fiprinca.0x90.eu/poc/
![cacache1](https://user-images.githubusercontent.com/16656956/36199310-24ccf66a-1171-11e8-8e29-ef0e41058943.png)
Thanks uBO :kiss:
![cacache2](https://user-images.githubusercontent.com/16656956/36199326-28e1e0da-1171-11e8-80dd-24677fd4934b.png)

**Comment by gorhill** [comment]
It's those 3 filters:


    ||flexytalk.net^
    ||cdn*.deadline.com^
    /?*&callback=callback_json_adblade*$script

But there is no point trying to escape such detection, and more generally, there is not much that can be done: any single filter list has it's own set of filters, and this can be used make a fingerprint of a filter list, to make a good guess of whether it's in use.

Blocking all scripts by default is the cure of such detection.

**Comment by overdodactyl** [comment]
Thanks for the info above Pants!

>get rid of the ... update app + symbol

You just wanted to get rid of the little icon when an update is available, right?

i.e. go from this

![update_available_icon](https://user-images.githubusercontent.com/33071129/36241462-6271ccb4-120e-11e8-8e48-195e50f60698.png)

to this

![no_icon](https://user-images.githubusercontent.com/33071129/36241532-d3177a4a-120e-11e8-9ddb-4403ee2b6497.png)


?

If so, this is what did it for me:

```css
#PanelUI-menu-button[badge-status="update-available"] > .toolbarbutton-badge-stack > .toolbarbutton-badge, 
#PanelUI-menu-button[badge-status="update-manual"] > .toolbarbutton-badge-stack > .toolbarbutton-badge, 
#PanelUI-menu-button[badge-status="update-restart"] > .toolbarbutton-badge-stack > .toolbarbutton-badge {
  display: none !important;
}
  
#PanelUI-menu-button[badge-status="update-available"], 
#PanelUI-menu-button[badge-status="update-manual"], 
#PanelUI-menu-button[badge-status="update-restart"]{
  list-style-image: url("chrome://browser/skin/menu.svg");
}
```

**Comment by Thorin-Oakenpants** [comment]
> You just wanted to get rid of the little icon when an update is available, right?

Yup. I find it annoying / distract... ooh look, a beer :beer: 

I know when I'm up to date etc with app releases, so I o not need this, plus I always have this (note haven't added your code yet, so not sure if you removed that, which I want to keep)
![meh](https://user-images.githubusercontent.com/16656956/36242702-3f67744c-1215-11e8-848a-5959d0addb30.png)

PS: that's a lot of code to remove one badge - do I really need all of it?

**Comment by Thorin-Oakenpants** [comment]
OK, cool, it removes the lot - that's fine by me :)

**Comment by overdodactyl** [comment]
>OK, cool, it removes the lot 

You sure I didn't remove too much? Not positive what ya wanted to keep.

>PS: that's a lot of code to remove one badge - do I really need all of it?

The first code chunk is to remove the green icon thingy.  The second is to replace the normal hamburger icon (when there's the update icon, it uses a different hamburger that's missing the upper right corner).

There's three id's for each because it appears they have different selectors for different updates (`update-available`, `update-manual`, `update-restart`)

**Comment by Thorin-Oakenpants** [comment]
I wouldn't mind keeping the menu item

**Comment by overdodactyl** [comment]
Sounds good...I updated my browser, so I'll have to wait until tomorrow to see it again/get back to ya

**Comment by Thorin-Oakenpants** [comment]
sorry for the OT guys & gals

@overdodactyl : my bad, I thought doing a manual check (Help>About FF) would be enough to trigger the badge etc. So I thought it was all hidden. Seems I was wrong, and I have to wait for the internal check to fire .. and the doorhanger notification.

the hamburger now has a bite taken out of it (see what I did there!), and menu item is still there, which is cool (because if I go to menu, it reminds me but is not a distraction)
![banana](https://user-images.githubusercontent.com/16656956/36255695-f598b348-1247-11e8-8f7b-11ee6db0034f.png)
